torchx/schedulers/ray_scheduler.py

```diff
@@ -6,56 +6,59 @@
 
 import logging
 import os
+import json
+import requests
+from tempfile import NamedTemporaryFile
+import dataclasses
 from dataclasses import dataclass, field
 from datetime import datetime
 from typing import Any, Dict, Iterable, List, Optional, Set, Type
+from .ray.ray_common import RayActor
 
-try:
-    import ray  # @manual # noqa: F401
 
-    _has_ray = True
-except ImportError:
-    _has_ray = False
+# try:
+import ray  # @manual # noqa: F401
+from ray.dashboard.modules.job.sdk import JobSubmissionClient
+from ray._private.job_manager import JobStatus
+from ray._private.job_manager import JobStatus
+from ray.autoscaler import sdk as ray_autoscaler_sdk 
+_has_ray = True
+
+# except ImportError:
+#     _has_ray = False
 
 from torchx.schedulers.api import AppDryRunInfo, DescribeAppResponse, Scheduler, Stream
 from torchx.schedulers.ids import make_unique
-from torchx.specs.api import AppDef, RunConfig, SchedulerBackend, macros, runopts
+from torchx.specs.api import AppDef, RunConfig, SchedulerBackend, AppState, macros, runopts 
 
 _logger: logging.Logger = logging.getLogger(__name__)
 
+def RayStatusResponseToTorchXAppState(jobstatus : JobStatus) -> AppState:
+    mapping = {
+        JobStatus.PENDING : AppState.PENDING,
+        JobStatus.RUNNING : AppState.RUNNING,
+        JobStatus.SUCCEEDED : AppState.SUCCEEDED,
+        JobStatus.FAILED : AppState.FAILED,
+        JobStatus.STOPPED : AppState.CANCELLED
+    }
 
-def has_ray() -> bool:
-    """Indicates whether Ray is installed in the current Python environment."""
-    return _has_ray
-
+    return mapping[jobstatus]
 
-@dataclass
-class RayActor:
-    """Describes an actor (a.k.a. role in TorchX terms).
 
-    Attributes:
-        name:
-            The name of the actor.
-        command:
-            The command that the actor should run as a subprocess.
-        env:
-            The environment variables to set before executing the command.
-        num_replicas:
-            The number of replicas (i.e. Ray actors) to run.
-        num_cpus:
-            The number of CPUs to allocate.
-        num_gpus:
-            The number of GPUs to allocate.
-    """
+class EnhancedJSONEncoder(json.JSONEncoder):
+        def default(self, o):
+            if dataclasses.is_dataclass(o):
+                return dataclasses.asdict(o)
+            return super().default(o)
 
-    name: str
-    command: str
-    env: Dict[str, str] = field(default_factory=dict)
-    num_replicas: int = 1
-    num_cpus: int = 1
-    num_gpus: int = 0
-    # TODO: memory_size, max_retries, retry_policy
+def serialize(actors : List[RayActor], output_filename='actors.json') -> None:
+    actors_json = json.dumps(actors, cls= EnhancedJSONEncoder)
+    with open(os.path.join("./ray", output_filename)) as tmp:
```

Comment:
This should use a temporary directory as I suggested in an earlier revision (i.e. `NamedTemporaryFile`). We should not pollute user's current working directory, in fact we might not even have write access to the directory, so this is prone to fail as well.